### DISCUSSION TITLE
Invoking storage API within Postgres

### DISCUSSION THREAD

**Discussion by bense30** [supabase/supabase#20717] [Questions] [answered]
I am trying to delete a specified folder within a bucket via a Postgres function. I have been able to use the below function to delete a specific object within the bucket, but when using the modified function (see second function) I get an error message saying "Object not found." Any help would be greatly appreciated. 

```
create or replace function delete_storage_object(bucket text, object text, out status int, out content text)
returns record
language 'plpgsql'
security definer
as $$
declare
  project_url text := '<YOURPROJECTURL>';
  service_role_key text := '<YOURSERVICEROLEKEY>'; --  full access needed
  url text := project_url||'/storage/v1/object/'||bucket||'/'||object;
begin
  select
      into status, content
           result.status::int, result.content::text
      FROM extensions.http((
    'DELETE',
    url,
    ARRAY[extensions.http_header('authorization','Bearer '||service_role_key)],
    NULL,
    NULL)::extensions.http_request) as result;
end;
$$;
```

Function for deleting a folder within the bucket:

```
create or replace function delete_storage_folder(bucket text, folder text, out status int, out content text)
returns record
language 'plpgsql'
security definer
as $$
declare
  project_url text := '<YOURPROJECTURL>';
  service_role_key text := '<YOURSERVICEROLEKEY>'; --  full access needed
  url text := project_url||'/storage/v1/object/'||bucket||'/'||folder;
begin
  select
      into status, content
           result.status::int, result.content::text
      FROM extensions.http((
    'DELETE',
    url,
    ARRAY[extensions.http_header('authorization','Bearer '||service_role_key)],
    NULL,
    NULL)::extensions.http_request) as result;
end;
$$;
```

I've also been unable to find the documentation for the Supabase Storage API, there is one online, but it is for V0 of the API.

**Comment by GaryAustin1** [comment, marked as answer]
There is no delete folder api.  You have to delete all the files by putting their paths in the body as an array or deleting them in a loop.
API https://supabase.github.io/storage/
My repository showing an automated method of cleaning up deleted files.
https://github.com/GaryAustin1/supa-file-helper
This should help you figure out your http or pg_net call to do multiple deletes if you want.

**Reply by bense30** [reply]
Hi Gary, Thank you for such a quick reply and for clearing that up for me. What I am ultimately trying to achieve is having all of the files uploaded by the user deleted when the user is deleted. It looks like supa-file-helper was created to help with something just like that, am I correct in thinking that?

Thank you again.

**Reply by GaryAustin1** [reply]
Yes.  I wanted a way to drop all the users notes and images of the notes.  There could be thousands.  So did not want to loop thru doing those api calls right when the user was deleted (auth has a time limit of like 3 seconds for the database call).  Also did not want to do individual deletes in a loop.

I have not re-looked at this since they changed to have a 2nd owner column so that might need adjusting, but I think is OK as is.  My project has been on hold so have not updated it this part yet.  

I would use it as a guide and note the risk if you do the same method of marking files for delete if they make changes.

**Reply by bense30** [reply]
It looks great to me! By "2nd owner column" are you referring to the owner_id column which is the owner UUID but in text form?

And lastly, out of curiosity, I was wondering why a solution might not be to query the storage.objects table for owner values that do not exist in auth.users and loop through those names deleting them? 

Thank you very much for your help.

**Reply by GaryAustin1** [reply]
Yes on the owner column.
The query to detect which owner column in one table is not in another seems pretty intense to be doing on a regular basis, but maybe there is a fast query to do that.  You don't control the storage schema so not sure if that column is indexed or not.  Auth.users id would be. I don't expect a lot of users being deleted 😄 so don't want to waste time on an expensive query each time.  You could look at that.  Certainly an efficient approach that does not rely on changing storage data is better.  

My main goal was to not delete at time of user deletion as I mentioned.  This solved it for me.

**Reply by bense30** [reply]
Makes a lot of sense. Thanks again for sharing.

**Reply by GaryAustin1** [reply]
I should have reread my repository....
![image](https://github.com/supabase/supabase/assets/54564956/23e1387b-5858-428a-979d-13e4bf443cfa)
The reason I could not even consider the auth compare method is storage used to have owner linked to auth.user id.  So to delete the user you had to delete storage objects OR like I did mark the owner null.  Then the user would delete.  That fk went away in Aug.  I've not really gone back to look at the complexity of comparing like you suggest.  I feel it is slow but I'm not going to dig into it any time soon.

**Reply by bense30** [reply]
Ah I see. Thanks for the extra clarity.
